### DISCUSSION TITLE
Tests not linking

### DISCUSSION THREAD

**Discussion by StefanGarnham** [rfennell/AzurePipelines#958] [Q&A] [answered]
Although I have found this to be useful, I have found the information on how the items in the extensions map to the actual objects in Azure DevOps.

I have created a test under a PBI which I expect to show under a build. It is not clear if tests can be shown in builds or if they are specifically for releases. I created a test plan and associated it with the build under my testing branch and this is not being pulled in either.

I would appreciate some assistance on getting this set up correctly.

Edited to add current handlebar template:

## Build {{buildDetails.buildNumber}}
* **Branch**: {{buildDetails.sourceBranch}}
* **Tags**: {{buildDetails.tags}}
* **Completed**: {{buildDetails.finishTime}}

# Work items ({{workItems.length}})
|ID | Description |
|---|---|
{{#forEach this.workItems}}
|[{{this.id}}]({{replace this.url "_apis/wit/workItems" "_workitems/edit"}})| {{{lookup this.fields 'System.Description'}}}|
{{/forEach}}

# Tests ({{tests.length}})
|ID | Description |
|---|---|
{{#forEach tests}}
|{{this.id}}| {{{lookup this.fields 'System.Description'}}}|
{{/forEach}}

**Comment by rfennell** [comment, marked as answer]
Tests are discovered in a variety of ways

| Where is the task run | What tests are loaded | Template object |
| - | - |- |
| Classic Build or YAML Build | The tests listed are any that have had their results published to the build e.g. VSTest Tests or similar.  | `tests`  |
| Classic Release (test included via artifacts) | Each artifact the release references is scanned in the same manner as a classic build (above). The test results are stored in two places | `tests` - the global list of unique tests run at any point in the pipeline <br> `builds[].tests` the tests specific to the given artifact |
| Classic Release | It is also possible to run and publish test results within a release stage, as opposed to in an artifact. Again these results are also stored in two places |  `tests` - the global list of unique tests run at any point in the pipeline <br> `releasetests` the tests specific to the release stages |

**Reply by StefanGarnham** [reply]
I am running a YAML build pipeline on a local branch to learn how to link everything together. I have a Test Plan, containing a test, which I have configured to the Build Pipeline I have created and Latest Build for the Build number. I have output the payload to the console and this is the resulting JSON. It contains no tests or build[].tests.

I can send the output JSON to you directly as I would prefer not to display potentially confidential information

**Reply by rfennell** [reply]
I think I understand the issue. 

There is no support for seeing any manual test (i.e the content of any runs from Azure DevOps Test Plans) data in the release notes. The task only for finds tests run and published using pipeline task such as VSTEST (or similar for other languages)

I guess it might be possible to make some extra API calls, to check the manual test run data. However, I not sure it would be that useful as the release notes would tend to be generated before anyone got around to running manual tests.

**Reply by StefanGarnham** [reply]
Would you please take this as a feature request? It would be great to be able to identify Test plans as we use these as part of our release notes.

**Reply by rfennell** [reply]
Certainly worth adding as a request. As I mentioned above not sure how well it will work, but it is worth some experimentation.

I have logged it #959

**Reply by rfennell** [reply]
#959 has been completed, it is now possible to reference manual test in a template

This all means I have had to add two new template objects

- manualtests - the array of manual Test Plan runs associated with any of the builds linked to the release
- manualTestConfigurations - the array of manual test configurations

Also, the template build array, used when the task is called in the release, has been extended with a manualtests property for each build.

These new objects can be used in templates to produce a simple list of test runs

```
## Manual Test Plans
| Run ID | Name | State | Total Tests | Passed Tests |
| --- | --- | --- | --- | --- |
{{#forEach manualTests}}
| [{{this.id}}]({{this.webAccessUrl}}) | {{this.name}} | {{this.state}} | {{this.totalTests}} | {{this.passedTests}} |
{{/forEach}}
```

If you wish to list all the individual tests you need to lookup the configuration for each test to get the configuration that was used.

```
## Manual Test Plans with test details
{{#forEach manualTests}}
### [{{this.id}}]({{this.webAccessUrl}}) {{this.name}} - {{this.state}}

| Test | Outcome | Configuration |
| - | - | - |
{{#forEach this.TestResults}}
| {{this.testCaseTitle}} | {{this.outcome}} | {{#with (lookup_a_test_configuration ../../manualTestConfigurations this.configuration.id)}} {{this.name}} {{/with}} |
{{/forEach}}
{{/forEach}}
```

These samples only show a few of the properties available on these new object. To see all the options use the `{{ json manualTests }}`
